src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,95 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
+        {
+            int length;
+            byte[] jsonBytes;
+            var sharedArrayPool = ArrayPool<byte>.Shared;
+            bool wasRented = false;
+            if (contentLength.HasValue)
+            {
+                length = (int)contentLength;
+                jsonBytes = sharedArrayPool.Rent(length);
+                wasRented = true;
+                await stream.ReadAsync(jsonBytes, 0, length);
+            }
+            else
+            {
+                using var ms = new MemoryStream();
+                await stream.CopyToAsync(ms).ConfigureAwait(false);
+                if (!ms.TryGetBuffer(out var streamBytes))
+                {
+                    return new GraphQLRequestDeserializationResult();
+                }
+                jsonBytes = streamBytes.Array;
```

Comment:
This isn't correct (it probably works but!). It would be better to pass around the span directly or `ReadOnlyMemory<byte>`. The Offset part of the `ArraySegment<byte>` is important as well.